Add failure detection with higher granularity for read replicas

psarna opened the pull request:
This series presents a mechanism of disqualifying a node from read operations if it looks down. It's based on estimating how long ago the target node was still responsive - an as that time gets longer than request's timeout, the chance for dropping the request increases.

Spin-off from #8098, which also claimed to be able to estimate reliably if a replica is overloaded, which is considerably harder to do properly.

psarna commented:
*ping*

psarna commented:
I'll also copy-paste grafana results from #8098, since they show the effects of the new failure detection:

> 1. A 3-node cluster is preloaded with some data, rf=3, cl=all
> 2. A quorum read workload starts - with custom USING TIMEOUT 1s CQL predicate; cl=quorum
> 3. One node is killed during the workload

> The red line in the middle indicates the point in which the cluster was upgraded to a version with the patches applied.

> What can be observed is that in the first case, right after one of the replicas is killed, we can observe a ~30s window of sharp increase in background reads along with accompanying read errors, as well as elevated number of digest reads, which is a result of speculative retry firing off.

> In the second case, background reads + read errors + digest reads are quickly translated to the new metrics - refused reads, which indicates that the node was disqualified as a candidate and the requests were rerouted to other nodes instead.

![with_and_without](https://user-images.githubusercontent.com/10433047/113270606-f3c0f600-92d9-11eb-8c45-d7c987a8f1c1.png)

psarna commented:
Also, refs https://github.com/scylladb/scylla/issues/3746#issuecomment-762203969 - once #3746 is fixed, the accidental backpressure of background reads will not work anymore.

psarna commented:
*ping*

psarna commented:
*ping²*

psarna commented:
ping³

nyh commented on service/storage_proxy.cc:
```diff
@@ -4601,6 +4636,16 @@ std::vector<gms::inet_address> storage_proxy::get_live_endpoints(keyspace& ks, c
+    auto& rs = ks.get_replication_strategy();
+    std::vector<gms::inet_address> eps = rs.get_natural_endpoints_without_node_being_replaced(token);
+    auto itend = boost::range::remove_if(eps, [this, timeout, &gossiper = gms::get_local_gossiper()] (const gms::inet_address& ep) {
+        return !gossiper.is_alive(ep) || !is_likely_to_respond_in_time(ep, timeout);
```
What if because of all this removal of candidates, we aren't left with enough candidates to fulfill CL? Maybe in this case we have to try those unlikely-to-respond nodes anyway? Or do we intentially want to return an error in this case - as a way to reduce the overload?

psarna replied:
Right, that's a very valid point. I suppose it may also depend what the client might want (which we could deduce from extended service level information, coming soon). Right now the code just prefers to drop the request fast even if it had a slim chance of succeeding, but it makes more sense to give some leeway if we're about to disqualify a node and it's crucial for CL to be fulfilled. That would render cl==ALL queries immune to this failure detector, but perhaps users sending cl==ALL requests are likely to want the query to succeed no matter what - otherwise, why set such a high consistency level.

I'll rethink the code to add a bypass for this disqualification mechanism if we're about to not fulfill the consistency level.

avikivity replied:
I think we should give each node a score (maybe true/false) and sort by score. This way we don't lose CL.

avikivity replied:
(but still prefer more-live replicas)

psarna commented on service/storage_proxy.cc:
```diff
@@ -3443,9 +3443,11 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
             return _proxy->query_mutations_locally(_schema, cmd, _partition_range, timeout, _trace_state);
         } else {
             tracing::trace(_trace_state, "read_mutation_data: sending a message to /{}", ep);
+            _proxy->mark_sent(ep, clock_type::now());
```
Hm, in case of detecting replica failures in order to potentially fail pending user requests it doesn't sound particularly bad to use the same resolution as for timeouts. After all, that's the resolution at which users already receive feedback if their queries timed out. I think that if the need for separate resolutions ever arises, it would be simple to just split then. After all, this typedef was named `clock_type`, which suggests a general-purpose clock for storage proxy if you're not interested in anything more specific. I think that if timeouts or any other modules want a de clock, they should define `timeout_clock_type` and use it.

psarna commented on service/storage_proxy.cc:
```diff
@@ -5254,6 +5262,25 @@ const db::hints::host_filter& storage_proxy::get_hints_host_filter() const {
+    // info should be reset in order to avoid having stale records.
+    if (info.last_sent + 5s < req_time) {
+        info.last_responded = last_seen_info::unknown;
+    }
```
Hm, it's been a month since I wrote all these cases, so I don't remember what I had in mind exactly, but I think I added this check after observing an anomaly when testing manually - e.g. something like that comes to mind:
1. A node gets extremely overloaded and stops responding to coordinators, but it's live enough not to get marked as DOWN
2. A coordinator tries communicating with it for a while and fails. It records that there were some communication attempts at time `T`
3. The coordinator stops communicating with the again, because clients' requests stopped coming
4. (1000 seconds pass)
5. The node regains health and is able to respond in time
6. The coordinator wants to communicate with the replica again - but it sees that it last sent the information over 1000 seconds ago and still hasn't received any response, so it qualifies the node as dead/overloaded, while it's perfectly healthy now

psarna replied:
The scenarios above are not supposed to be likely because the database of recent timestamps is reset each time a node is reported as added/removed/dead/alive, but I think it may still happen

psarna commented on service/storage_proxy.cc:
```diff
@@ -4593,6 +4595,39 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
+        if (last_seen_ep->last_responded == last_seen_info::unknown || last_attempt > time_left) {
+            return true;
+        }
+        auto time_without_response = last_seen_ep->last_sent - last_seen_ep->last_responded;
```
After reading this code after a month, I'm confused... doesn't sounds like a right calculation indeed. It gets bogus results each time stop sending for a couple of seconds, and then send again but then check before the replica has a chance to respond. I'll try to figure out what I had in mind and either provide a meaningful explanation or rewrite the whole thing. "Time without response" is literally what you described in the comment, not what is calculated here.

psarna replied:
Oh wait, now, after I responded to your other comment, I remember more. I'm not really interested in `time_without_response` as in "how long ago we received a response from a node", but rather "how long did it take last time for this node to respond" - the name is confusing and unfortunate indeed. The problem is that I also want to estimate this value with as few bytes used and cpu cycles wasted, so it's only a very rough estimate. This formula is supposed to work well if the system is under constant load - and so we also very frequently try to contact all the replicas. If we don't contact the replicas frequently, the other mechanism (which you also asked about in a previous comment) checks if the information is not too stale - and if it is, it's dropped. With these two checks in place, it's possible to roughly guess if the node is responsive or not, but at the same time do not keep track of latencies of each request - because that would have been much costlier to maintain than a single pair of timestamps per endpoint. So it's entirely possible (and expected) that `last_sent` and `last_responded` do not actually come from the same request, but if the coordinator node is sufficiently busy, it's still enough to examine the values to detect if a target node might be dead - which would be evident after seeing lots of sending attempts, but no responses.

psarna replied:
Still, I'm not fully utilizing the information I have. Instead of just collecting `last_sent` and `last_responded` from potentially unrelated requests, it's easy enough to register the actual time it took for a full round trip when registering a response time - since when the response is registered, we still know when it was sent. Using that would simplify these questionable calculations and hopefully make the review clearer as well. I'll prepare a better v2.

psarna commented on service/storage_proxy.cc:
```diff
@@ -4593,6 +4595,39 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
+            static constexpr int granularity = 10'000;
+            static thread_local std::uniform_int_distribution dice(0, granularity);
+            // There's always a small chance of probing the replica in case it came back to life.
+            int fail_chance = std::min<long>((time_without_response.count() - time_left.count()) * granularity / time_left.count(), granularity - 1);
```
*Assuming* that `time_without_response` is calculated correctly and simply means how long ago a node was responsive (taking your previous comment it looks like it's not the case, and I'll fix that), then what I had in mind was to linearly increase the probability of qualifying the node as dead/overloaded, the more we cross the deadline, reaching 100% at 2*timeout. So if the timeout for a request is set to 100ms and last time a node was able to respond it took it around `X` ms, there's `(100-X)%` chance the node should be omitted, which translates to 0 if the timeout deadline was not crossed yet and approaching 100% if the last time a node responded it took it more than 2*timeout to do so.

psarna commented:
I pushed another iteration by the end of the day, but please do not review yet. I reworked the estimation calculations and added a fail-safe for the case when there are not enough replicas to satisfy the consistency level, but I haven't had time to run my test scenarios and update the docs yet.

psarna commented:
v2:
 - reworked the calculations: the values gathered now are:
   - timestamp of a first sending attempt that didn't get its response
   - timestamp of the most recent sending attempt
   - round-trip time of the last successful request
 - added a fail-safe for the case when disqualifying a node also means that we won't be able to fulfill the consistency level - in that case, it's preferred to at least try contacting the replica, even if it's likely to be dead
 - amended the docs accordingly

Tests: manual, same scenarios as before, but for ONE, QUORUM and ALL consistency levels; testing with ALL allows checking if the fail-safe mechanism works

psarna commented:
Example microbenchmark results below (`perf_simple_query --smp 1` ran multiple times).
master:
```
43087.77 tps ( 87.2 allocs/op,  20.2 tasks/op,   85885 insns/op)
43292.30 tps ( 87.2 allocs/op,  20.2 tasks/op,   85870 insns/op)
43154.16 tps ( 87.2 allocs/op,  20.2 tasks/op,   85871 insns/op)
43158.22 tps ( 87.2 allocs/op,  20.2 tasks/op,   85870 insns/op)
42427.62 tps ( 87.2 allocs/op,  20.2 tasks/op,   85868 insns/op)

median 43154.16 tps ( 87.2 allocs/op,  20.2 tasks/op,   85871 insns/op)
median absolute deviation: 66.39
maximum: 43292.30
minimum: 42427.62
```
this-series:
```
42220.20 tps ( 87.2 allocs/op,  20.2 tasks/op,   86184 insns/op)
42585.07 tps ( 87.2 allocs/op,  20.2 tasks/op,   86189 insns/op)
43336.23 tps ( 87.2 allocs/op,  20.2 tasks/op,   86227 insns/op)
42968.22 tps ( 87.2 allocs/op,  20.2 tasks/op,   86177 insns/op)
42978.42 tps ( 87.2 allocs/op,  20.2 tasks/op,   86196 insns/op)

median 42968.22 tps ( 87.2 allocs/op,  20.2 tasks/op,   86177 insns/op)
median absolute deviation: 368.01
maximum: 43336.23
minimum: 42220.20
```
Sounds like any change between master and this series is within the error margin.

psarna commented:
*ping for v2*

avikivity commented:
Usually the instructions counter is fairly accurate. It's unreasonable large, are you testing dev mode?

For stablest results, use perf_simple_query --smp 1 --task-quota-ms 10

psarna commented:
I think it was release mode, but I have inline threshold set to 0 in order to avoid clang hanging, which may influence results. In any case, I'll just rerun with task-quota-ms 10 and see.

psarna commented:
(github issued an update because I pushed a rebased version, but I haven't rerun the perf test yet)

psarna commented:
Bumped the inline threshold to 300 (600 is already enough to hang the compilation of row_level.cc, so I gave up higher numbers), here are the results:
```
Running test with config: {partitions=10000, concurrency=100, mode=read, frontend=cql, query_single_key=no, counters=no}
Creating 10000 partitions...
76410.47 tps ( 87.1 allocs/op,  20.1 tasks/op,   62250 insns/op)
75873.35 tps ( 87.1 allocs/op,  20.1 tasks/op,   62176 insns/op)
78467.92 tps ( 87.1 allocs/op,  20.1 tasks/op,   62209 insns/op)
80713.69 tps ( 87.1 allocs/op,  20.1 tasks/op,   62164 insns/op)
80561.72 tps ( 87.1 allocs/op,  20.1 tasks/op,   62174 insns/op)

median 78467.92 tps ( 87.1 allocs/op,  20.1 tasks/op,   62209 insns/op)
median absolute deviation: 2093.80
maximum: 80713.69
minimum: 75873.35
```
Is the instruction count still huge? I don't really have a reference

avikivity commented:
The values I remember are 55k insn/query (that's high too).

But if it doesn't compile, how can we merge it?

psarna commented:
The master branch doesn't compile for me since a couple of months - it's not the matter of this patch at all.

denesb commented:
On Thu, 2021-05-27 at 07:38 -0700, Avi Kivity wrote:
> Usually the instructions counter is fairly accurate. It's
> unreasonable large, are you testing dev mode?
> For stablest results, use perf_simple_query --smp 1 --task-quota-ms
> 10
Should we make this the default, to avoid everybody running this with
the wrong arguments out-of-the-box?

> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

psarna commented:
*review ping*

psarna commented:
edit: I'll get acquainted with the new failure detector first, since it's merged, and then go back to this change - maybe it's no longer needed

psarna commented:
edit2: no clash - the new direct failure detector still uses rather high granularity with default timeout of 20s, while this series targets requests with custom timeouts. As an example, ones with timeout set to 200ms will qualify nodes are dead considerably faster (~400ms), which makes the coordinator much less prone to getting overloaded with hanging background reads.

psarna commented:
*review ping*

psarna commented:
*review ping*

avikivity commented:
Can the mechanism disqualify two nodes?

In general, we should disqualify at most (live nodes - requires_for_cl) nodes, otherwise we risk a false positive causing us to fail CL.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -5388,6 +5400,25 @@ future<> storage_proxy::wait_for_hints_to_be_replayed(utils::UUID operation_id,
 }
 
+void storage_proxy::mark_responded(const gms::inet_address& ep, clock_type::duration rtt) {
+    last_seen_info& info = _last_seen[ep];
```
We really need a node_ptr to avoid all those hash tables everywhere indexing the same thing.

Please provide before/after perf_simple_query instruction count to see how much this costs.

psarna replied:
https://github.com/scylladb/scylla/pull/8393#issuecomment-834477194

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4610,6 +4612,45 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
+    }
+    if (fbu::is_me(ep)) {
+        return true;
+    }
```
I could be overloaded.

psarna replied:
I never reached such a scenario in my tests, but I think that if the coordinator node is already overloaded, it should either start returning `OverloadedExeption` to clients or creating backpressure with the admission semaphore, and ideally never even reach a point when it needs to exclude itself as a replica in order to serve other requests. Still, maybe the assumption should be revisited, I'll try to figure out how to reach such scenario (one obvious way would be to use a coordinator with really slow hard drives).

avikivity replied:
It could be overloaded as a replica but not as a coordinator (disk is overloaded but not cpu).

avikivity replied:
We can just drop this check, no? And gather the info for is_me().

psarna reviewed (commented):
@avikivity I addressed the CL problem in v2 changelog: https://github.com/scylladb/scylla/pull/8393#issuecomment-832675800
tldr: if disqualifying a node would mean not being able to fulfill the CL, we contact the node anyway instead

psarna commented:
Also: I'm on vacation next week and there are merge conflicts already (and at a rate in which optimizations are applied to this area, new conflicts are guaranteed to accumulate), so I'll need to do a proper rebase of this series once I'm back.

avikivity commented:
> Also: I'm on vacation next week and there are merge conflicts already (and at a rate in which optimizations are applied to this area, new conflicts are guaranteed to accumulate), so I'll need to do a proper rebase of this series once I'm back.

The trick is to find problems with the optimizations so that they are delayed.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4618,6 +4659,24 @@ inet_address_vector_replica_set storage_proxy::get_live_endpoints(keyspace& ks,
+    });
+    // In the unlikely case when too many endpoints were disualified and it's impossible to fulfill
+    // the consistency level, bring back the endpoints which were "not likely to respond in time"
+    // and only remove the dead ones
```
Actually, it occurs to me that the original behavior was also useful. If we think that the nodes are unlikely to respond in time, why make the attempt?

But, this requires that we have great trust in this "unlikely to respond in time".

nyh replied:
If the request has CL=ALL (for example) and a timeout of 10 seconds, why would you want to return an error soon instead of waiting 10 seconds and return a timeout then? You know one of the replicas is dead, but maybe it will come back in 9 seconds, and we'll be able to make the 10 second deadline? Where's the benefit of returning sooner?

The benefit we do have is when CL<ALL so speculative retry generates a lot of "background reads" which pile up without limit. So this is what this patch tries to avoid. It's not about timing out sooner for user requests.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4618,6 +4659,24 @@ inet_address_vector_replica_set storage_proxy::get_live_endpoints(keyspace& ks,
+    if (!db::is_sufficient_live_nodes(cl, ks, eps.begin(), itend)) [[unlikely]] {
+        itend = std::remove_if(itend, eps.end(), [&gossiper = gms::get_local_gossiper()] (const gms::inet_address& ep) {
+            return !gossiper.is_alive(ep);
+        });
```
Please check with perf_simple_query. is_sufficient_live_nodes() looks quite large.

psarna commented:
rebased to get rid of the conflicts

nyh commented on service/storage_proxy.cc:
```diff
@@ -3465,9 +3465,12 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
             tracing::trace(_trace_state, "read_mutation_data: sending a message to /{}", ep);
-            return _proxy->_messaging.send_read_mutation_data(netw::messaging_service::msg_addr{ep, 0}, timeout, *cmd, _partition_range).then([this, ep](rpc::tuple<reconcilable_result, rpc::optional<cache_temperature>> result_and_hit_rate) {
+            clock_type::time_point sent_at = clock_type::now();
+            _proxy->mark_sent(ep, sent_at);
```
If you plan to only remember when you sent *read* requests and recieved responses for *read* requests, then you should call it mark_sent_read() or something. But I wonder - if we're looking for down nodes, not "slow" nodes (and the commit message could use some clarification on this), why only look at read messages? Any received message is an indication that the node is not down. Maybe we can do this mark_sent/mark_responded in some lower level function, and not in multiple read/write message implementations?

psarna replied:
Yeah, it may make sense. On the other hand, now the extra overhead is reduced to the only interested party - the read mechanism. I'll look for a way to push it to a lower layer though.

nyh commented on service/storage_proxy.cc:
```diff
@@ -3465,9 +3465,12 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
+            return _proxy->_messaging.send_read_mutation_data(netw::messaging_service::msg_addr{ep, 0}, timeout, *cmd, _partition_range).then([this, ep, sent_at](rpc::tuple<reconcilable_result, rpc::optional<cache_temperature>> result_and_hit_rate) {
                 auto&& [result, hit_rate] = result_and_hit_rate;
                 tracing::trace(_trace_state, "read_mutation_data: got response from /{}", ep);
+                _proxy->mark_responded(ep, clock_type::now() - sent_at);
```
I'm still trying to figure out the theory of what it is you're trying to detect - a node being down or being slow :-(
To detect down nodes I'd expect you should remember *when* the node last responded - i.e., now() - not the latency now() - sent_at as you do now... But even with now(), it's not trivial how exactly you detect a node being down. Maybe you can write some of the "algorithm" in the commit message.

psarna replied:
Right, I think it's an artifact from the time when this series was both about detecting dead and overloaded nodes. This failure detector shouldn't be interested in how much time it took for a node to respond, as long as it responded at all. I'll go over this code one more time.

nyh commented on docs/design-notes/qualifying-replicas-for-reads.md:
```diff
@@ -0,0 +1,83 @@
+# Qualifying replicas for read operations
+
+## Intro
```
Is this document supposed to cover all the aspects of choosing replicas?
If so it should also mention heat-weighted-load-balancing, snitches (not just "dynamicsnitch"), etc.
Or, if it's supposed to be a more specific document, the intro should explain it only covers part of the choosing-replicas-for-read story and that other parts are documented elsewhere.

nyh replied:
By the way, can you move this patch to be first instead of last? It can help reviewers understand the theory behind the code they'll see.

psarna replied:
Sure, I'll move it as the first patch. And I mean it to eventually cover all aspects, but it's out of scope for this series - I just hope that one day it will be filled with more information about other mechanisms that Scylla has.

nyh commented on docs/design-notes/qualifying-replicas-for-reads.md:
```diff
@@ -0,0 +1,83 @@
+mechanism with high granularity is implemented as follows:
+ * each time a read request is sent to a replica, this attempt's timestamp
+   is recorded
+ * first timestamp of a streak of unresponded requests is recorded separately
```
I'm not sure what this streak means, or how you can calculate it, maybe you can explain this better. Imagine in times 1,2,3...100 you send 100 requests, and until now requests 2 through 70 were responded. What is the "first timestamp of a streak of unresponded requests" - is it time 1? 71? or what?

psarna replied:
(I hope I still remember and understand my original idea...) 71 - what I meant to express is that if there's a series of requests that didn't get any response, I remember the timestamp of the first one. Here's an example for time points 1 to 8:
1. send A
2. send B
3. recv A
4. recv B
5. send C <---- this is the timestamp I'm interested in (from the point of view of time point 8. below); that's when we stopped getting any responses
6. send D
7. send E
8. \<now>

nyh commented on docs/design-notes/qualifying-replicas-for-reads.md:
```diff
@@ -0,0 +1,83 @@
+round-trip is a cheap, rough implementation of a failure detector - if the node hasn't
+responded for more than X milliseconds since the previous request and the current request
+is going to time out in X milliseconds, the target node gets more and more likely to be
+qualified as dead.
```
Again I think there's a confusion between "dead" and a node which because of some overload problem now has latency X for each request but it still very much alive. We should be clear which problem we're tackling.

psarna replied:
Yes, when I started writing this series and this document, I had both failure detection and dynamic snitch in mind, but then the scope got reduced to just failure detection. So ultimately I want to be able to detect nodes which are either dead or look dead (e.g. due to long network partition).

psarna replied:
And my intention with tying the detection with timeouts is to make the failure detector more strict for requests with short timeouts. The idea is that when a workload is full of requests with 200ms timeouts, it makes sense to have a more strict failure detection in order to minimize the time wasted on sending requests to dead replicas. On the other hand, long-standing requests with multi-second timeouts can be more lenient and give more nodes a chance to respond.

psarna commented:
I pushed another rebase, but it's not a new iteration yet - only the order of patches was changed. Nadav's comments are still to be applied in the new iteration, which will come soon.

psarna commented:
v3:
 - moved the detector one level down, the timestamps are now gathered in the messaging service, so all two-way messages contribute to updating the values
 - removed last remnants of gathering round-trip times, which were there in initial iterations, back when the series was both about failure detection and overload detection - now the detector only cares about detecting dead nodes, so round-trip times are irrelevant
 - updated docs accordingly
 
Tests: unit(release), manual(killing 1 node from a 3-node cluster during quorum reads)

psarna commented:
weekly ping

nyh commented on message/messaging_service.cc:
```diff
@@ -250,6 +250,17 @@ rpc::stats messaging_service::shard_info::get_stats() const {
 }
 
+void last_seen_info::mark_sent(clock_type::time_point req_time) {
+    last_attempt = req_time;
```
Ok, I think I understood last_attempt. Resolving this thread.

psarna commented on docs/design-notes/qualifying-replicas-for-reads.md:
```diff
@@ -0,0 +1,90 @@
+At first, we compute how much time left (`TL`) the request still has until
+it reaches its deadline and times out. From the last seen information
+we fetch the last time without any response: `TWR`.
+If `TL` is larger than `TWR`, then the node is likely alive, so the request
```
I'm not entirely convinced about this idea either - I just liked the fact that **if we assume** that interactive workloads use small timeout values and batch workloads use large timeout values, we'll get more aggressive failure detection for interactive workloads. That's because I treat a small timeout as an SLA expectation - "I need this query to finish in 200ms or I don't care about the result anymore". On the other hand, long timeout values suggest that the request can take its time to finish, and it's more important to finish it eventually that to kill it and have better latency for other similar requests.

On the other hand, we already have another mechanism for specifying the workload type - users can set it up via service levels. So perhaps the more aggressive failure detection can only be activated for interactive workloads.

Finally, if we don't use `TL`, do you have another suggestion what to base the computations on? We still need a value to decide what's the sensitivity of this failure detector. Some alternatives are
 - a hardcoded value that we deem reasonable, e.g. "5 seconds"
 - a configuration variable (perhaps per-service-level)

Suggestions/ideas?

psarna replied:
*discussion ping*

psarna commented on service/storage_proxy.cc:
```diff
@@ -4639,6 +4641,43 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
+    // - time without getting any response since the last time the coordinator started trying to communicate
+    //   with the replica
+    if (time_without_response > time_left) {
+        static constexpr int granularity = 10'000;
```
Precisely - it's an arbitrary value to avoid floating point arithmetics, since we don't need such high precision anyway.

psarna commented:
not going to be merged, closing
